doc/dev/error-categorization.md

```diff
@@ -0,0 +1,111 @@
+we were facing many duplicated errors because too many existed and developers simply cannot know each
+and every code.
+
+So if you feel the need for a new category you should first ask yourself the following question:
+> Can I react differently if I would know every error X in category Y?
+
+If not it will most likely fit into an existing category. Categories are leaf based, so you cannot
+put an error into a node such as `Permanent errors` (see below).
+
+## Error categorization Guideline
+
+Now we will investigate each category in more detail and when to put an error/warning in there.
+
+The current structure looks like this:
+- Permanent errors ("01000")
+  - Resource ("01100")
+  - Parsing ("01200")
+  - Installation ("01300")
+  - Logical ("01400")
+- Conflict ("02000")
+- Timeout ("03000")
+- Validation ("04000")
+  - Syntactic ("04100")
+  - Semantic ("04200")
+### Permanent errors ("01000")
+
+The main category `Permanent Errors` refer to such errors which cannot be fixed by retry 
+at all. `Permanent Errors` are subdivided into
+- Resource
+- Parsing
+- Installation
+- Logical
+#### Resource ("01100")
+
+`Resource Errors` are all kinds of errors which are either permission related, existence related
+or missing resources such as memory, RAM, etc.
+Examples are missing files, insufficient permissions or out of memory.
+This category forces users or applications to provide additional resources, create a missing file/ directory etc
+or provide the relevant permissions for Elektra to work.
+
+#### Parsing ("01200")
+
+`Parsing Errors` are reserved for parsers for xml/yaml/ini etc. and indicate errors which
+were investigated during parsing. Such errors show unexpected encounters like missing line endings,
+illegal characters or wrong encodings.
+Users or applications will have to sanitize the input and retry again.
+
+#### Installation ("01300")
+
+`Installation Errors` are those errors which are related to Elektra's installation such as
+wrong plugin names, missing backends, initialization errors, misconfiguration of Elektra etc.
+Users will have to reconfigure, reinstall, recompile (with other settings) Elektra in order to
+get rid of this error. Applications will most likely not be able to handle such errors by themselves.
+
+#### Logical ("01400")
+
+`Logical Errors` indicate a bug in Elektra such as internal errors, assertion failures or errors
+which should not happen such as going into a `default` branch when you are assured that all cases
+are covered. Usually such errors come with a message to report such failures to Elektra's bugtracker.
+Applications cannot handle such errors themselves other than reporting it to Elektra and hope
+that the developers investigate and fix the issue.
+
+### Conflict ("02000")
+
+`Conflict Errors` are errors which indicate temporary problems which can be handled by handled by
+updating the current state. Examples are the need for calling `kdbGet` before `kdbSet`, or pulling changes
```

Comment:
What is "the current state"? A configuration file is technically also part of the current state, but that obviously doesn't fit this case.

Reply:
@markus2330 requested this category and I am also not 100% sure. I assume that it is an error where you try to modify another state but see that it differs from what you expect (git push, file changed externally, kdb get before kdb set) .. does this make more sense to you?

Reply:
I would say "an error where the current state is incompatible with the attempted operation".

The incompatibility is what causes the problem. Whether this incompatibility is temporary or not or what needs to change shouldn't really matter I think.